Error estimates for approx quantiles, better PDF plots.

patrick-schultz opened the pull request:
Add an a postiori error estimate for the approximate cdf aggregator. Use in new pdf plotting method than makes no assumption about the smoothness of the true distribution.

catoverdrive commented on hail/src/main/scala/is/hail/annotations/aggregators/RegionValueApproxCDFAggregator.scala:
```diff
@@ -346,14 +349,26 @@ class ApproxCDFCombiner[@specialized(Int, Long, Float, Double) T: ClassTag : Ord
       lvl += 1
     }
 
+    lvl = 0
```
it looks like you're adding both compaction counts to `mergedCompactionCounts(0)` twice here; is that intentional?

patrick-schultz replied:
Oops, no, good catch!

catoverdrive commented on hail/src/main/scala/is/hail/annotations/aggregators/RegionValueApproxCDFAggregator.scala:
```diff
@@ -248,18 +252,15 @@ class ApproxCDFCombiner[@specialized(Int, Long, Float, Double) T: ClassTag : Ord
-  }
-
-  def compactLevel(level: Int, shiftLowerLevels: Boolean) {
+  def compactLevel(level: Int, shiftLowerLevels: Boolean = true, dummy: T = helper.dummyValue) {
```
What's the purpose of these dummy variables? I don't see anywhere these functions are overloaded?

patrick-schultz replied:
Without the dummy variable, this function would have no `T` in its signature, which means Scala would refuse to generate specialized methods. But we very much want the implementation of `_compactLevel` to be specialized for each concrete type, to avoid boxing, and unspecialized methods always call unspecialized versions of any called methods. There were just a few spots I missed adding dummy parameters in the original PR.

catoverdrive replied:
ah, cool!

catoverdrive commented on hail/python/hail/expr/aggregators/aggregators.py:
```diff
@@ -235,7 +235,7 @@ def approx_cdf(expr, k=100):
         Struct containing `values` and `ranks` arrays.
     """
-    return _agg_func('ApproxCDF', [expr], tstruct(values=tarray(expr.dtype), ranks=tarray(tint64)), constructor_args=[k])
+    return _agg_func('ApproxCDF', [expr], tstruct(values=tarray(expr.dtype), ranks=tarray(tint64), _compaction_counts=tarray(tint32)), constructor_args=[k])
```
I don't super like that we've got undocumented fields in the return type; would it be reasonable to document?

patrick-schultz replied:
What we really want is for the `ApproxCDF` aggregator to return a custom hail type, that hides implementation details like compaction counts, and supports methods like estimating quantiles or getting error bounds. But that should probably wait until physical types are more developed. Or maybe it's reasonable to do now; I haven't looked into it.

I can add documentation that just says this is an implementation detail, so users aren't confused by it.

catoverdrive commented on hail/python/hail/expr/functions.py:
```diff
@@ -48,6 +49,17 @@ def _quantile_from_cdf(cdf, q):
+    Error of any one quantile estimation is less than returned bound with
+    probability 1-'confidence'.
+    """
+    s = range(0, len(cdf._compaction_counts)).map(lambda i: cdf._compaction_counts[i] << (2*i)).sum
```
does the `.sum` notation actually work?

catoverdrive replied:
Also, I know most of these functions exist directly in the namespace, but I think it's nice to prefix with `hl.` anyways to make it clear that we're operating on expressions. We also may want to bind the `cdf` expression so we're not potentially computing the aggregation twice, although if it's not bound outside the scope of this function then we'd need another computation to get the actual cdf anyways?

patrick-schultz commented:
> Some small comments, but I'm also not sure I totally understand what's going on with the new pdf function.

I can explain in person, it's pretty cool (I think). That might also help me figure out something to say in the documentation, though since it's experimental I don't think it needs complete documentation for now.

> Would it be possible to add some python tests for the new stuff?

Do you think tests that just check that the `_compaction_counts` field exists are worth having? Beyond that, everything is non-deterministic and hard to test, and I don't know how to test plotting methods either. But I've been playing with them and they are working extremely well.

catoverdrive commented:
I think that would probably be fine---I just got worried since `_error_from_cdf` had some non-working code and wasn't being called anywhere explicitly (it's imported in `plots`, but not used in the error calculation)

patrick-schultz commented:
I moved the changes to the aggregator into #6076, and added error bounds that apply to all quantiles simultaneously, which is what you really want for the pdf.
